docs/migrating.rst

```diff
@@ -0,0 +1,275 @@
+######################################
+Migrating from Python 2 version of NDB
+######################################
+
+While every attempt has been made to keep compatibility with the previous
+version of `ndb`, there are fundamental differences at the platform level,
+which have made necessary in some cases to depart from the original
+implementation, and sometimes even to remove existing functionality
+altogether.
+
+One of the main objectives of this rewrite was to enable `ndb` for use in any
+Python environment, not just Google App Engine. As a result, many of the `ndb`
+APIs that relied on GAE environment and runtime variables, resources, and
+legacy APIs have been dropped.
+
+Aside from this, there are many differences between the Datastore APIs
+provided by GAE and those provided by the newer Google Cloud Platform. These
+differences have required some code and API changes as well.
+
+Finally, in many cases, new features of Python 3 have eliminated the need for
+some code, particularly from the old `utils` module.
+
+If you are migrating code, these changes can generate some confusion. This
+document will cover the most common migration issues.
+
+Setting up a connection
+=======================
+
+The most important difference from the previous `ndb` version, is that the new
+`ndb` requires the use of a client to set up a runtime context for a project.
+This is necessary because `ndb` can now be used in any Python environment, so
+we can no longer assume it's running in the context of a GAE request.
+
+The `ndb` client uses ``google.auth`` for authentication, which is how APIs in
+Google Cloud Platform work. The client can take a `credentials` parameter or
+get the credentials using the `GOOGLE_APPLCATION_CREDENTIALS` environment
+variable, which is the recommended option.
+
+After instantiating a client, it's necessary to establish a runtime context,
+using the ``Client.context`` method. All interactions with the database must
+be within the context obtained from this call::
+
+    from google.cloud import ndb
+
+    client = ndb.Client()
+
+    with client.context() as context:
+        do_something_with_ndb()
+
+The context is not thread safe, so for threaded applications, you need to
+generate one context per thread. This is particularly important for web
+applications, where the best practice would be to generate a context per
+request.
```

Comment:
This is confusing. If you have a web application, where the best practice would be to generate one context per request, and you need multiple threads inside that request, what should you do?

If the answer is "generate multiple contexts" then maybe something like this?

"In a web application, a typical best practice is to generate one context per request. However, a context object only applies to a single thread; in the case that multiple threads are used for a single request, a new context should be generated for every thread that will use the NDB library."